atom/browser/api/atom_api_app.cc

```diff
@@ -1162,6 +1164,33 @@ v8::Local<v8::Value> App::GetGPUFeatureStatus(v8::Isolate* isolate) {
   return mate::ConvertToV8(isolate, status ? *status : temp);
 }
 
+util::Promise* App::GetGPUInfo(v8::Isolate* isolate,
+                               const std::string& info_type) {
+  if (info_type != "available" && info_type != "complete") {
+    const auto& promise = new util::Promise(isolate);
+    promise->Reject();
+    return promise;
+  }
+  const auto gpu_data_manager = content::GpuDataManagerImpl::GetInstance();
+  if (!gpu_data_manager->GpuAccessAllowed(nullptr)) {
+    const auto& promise = new util::Promise(isolate);
+    promise->Reject();
+    return promise;
+  }
+
+  if (info_type == "complete") {
+    // In this case, we create an object on the heap and GPUInfoManager is
+    // responsible for deleting it
+    GPUInfoManager* info_mgr = new GPUInfoManager(isolate);
```

Comment:
Is there something like `enable_shared_from_this` for chromium, so that I do not have to do this memory management dance?

Reply:
The `GPUInfoManager` can be a `base::Singleton<>`, which takes ownership of the `util::Promise`. This also allows the class to be moved outside the api layer.